src/executors/executor_sycl.hpp

```diff
@@ -230,6 +230,88 @@ Executor<PolicyHandler<codeplay_policy>>::execute(
            Gemm_type>::local_memory_size)};
 }
 
+/* Utility function used by the ReductionPartialRows specialization */
+template <typename operator_t, int ClSize, int WgSize, int WorkPerItem,
+          typename element_t, typename input_t, typename output_t,
+          typename index_t, typename queue_t>
+static inline cl::sycl::event launch_row_reduction_step(
+    queue_t queue, input_t& in, output_t& out, index_t rows, index_t cols,
+    index_t group_count_cols, index_t local_memory_size,
+    index_t num_compute_units) {
+  ReductionPartialRows<operator_t, input_t, output_t, ClSize, WgSize,
+                       WorkPerItem, element_t>
+      reduction_step(in, out, rows, cols, group_count_cols);
+  auto step_range = reduction_step.get_nd_range(num_compute_units);
+  return execute_tree<using_local_memory::enabled>(
+      queue, reduction_step, step_range.get_local_range()[0],
+      step_range.get_global_range()[0], local_memory_size);
+}
+
+/* ReductionPartialRows */
+template <>
+template <typename operator_t, typename input_t, typename output_t, int ClSize,
+          int WgSize, int WorkPerItem, typename element_t>
+inline typename codeplay_policy::event_t
+Executor<PolicyHandler<codeplay_policy>>::execute(
+    Reduction<operator_t, input_t, output_t, ClSize, WgSize, WorkPerItem,
+              element_t, static_cast<int>(Reduction_t::partial_rows)>
+        reduction_wrapper) {
+  using index_t = typename std::make_signed<typename input_t::index_t>::type;
+  using params_t = blas::ReductionRows_Params<index_t, element_t, ClSize,
+                                              WgSize, WorkPerItem>;
+
+  /* Extract data from the reduction wrapper */
+  const index_t rows_ = reduction_wrapper.rows_,
+                cols_ = reduction_wrapper.cols_;
+  input_t &in_ = reduction_wrapper.in_, out_ = reduction_wrapper.out_;
+
+  const index_t num_compute_units = policy_handler_.get_num_compute_units();
+
+  /* Choose at run-time whether to do a one-step or two-step reduction */
+  const bool do_first_step =
+      cols_ > params_t::work_group_cols * params_t::work_group_cols;
+  // TODO: find out in which cases it is better to use 2-step reduction
```

Comment:
Obligatory "Does this need a Jira?"

Reply:
It will be part of the optimization of tall and skinny gemm, I guess. Ideally there should be a benchmark for the reduction but the current benchmarks only measure the performance of the API and not the underlying kernels.

Reply:
I've created a JIRA issue.